src/ert/storage/local_ensemble.py

```diff
@@ -885,73 +909,118 @@ def get_observations_and_responses(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
                     )
-                )
-                .drop(response_cls.primary_key)
-                .rename({"__tmp_index_key__": "index"})
-            )
 
-            first_columns = [
-                "response_key",
-                "index",
-                "observation_key",
-                "observations",
-                "std",
-            ]
-            joined = joined.select(
-                first_columns + [c for c in joined.columns if c not in first_columns]
-            )
+                    # Filter out irrelevant data before reading into memory
+                    for col, observed_values in observed_cols.items():
+                        my_responses = my_responses.filter(
+                            polars.col(col).is_in(observed_values)
+                        )
+
+                    pivoted = my_responses.collect().pivot(
+                        on="realization",
+                        index=["response_key", *response_cls.primary_key],
+                        values="values",
+                        aggregate_function="mean",
+                    )
+
+                    if pivoted.is_empty():
+                        joined = observations_for_type.with_columns(
+                            polars.Series(
+                                str(real),
+                                [np.nan] * len(observations_for_type),
+                                dtype=polars.Float32,
+                            )
+                        )
+                    elif "time" in pivoted:
+                        joined = observations_for_type.join_asof(
+                            pivoted,
+                            by=["response_key", *response_cls.primary_key],
+                            on="time",
+                            tolerance="1s",
+                        )
+                    else:
+                        joined = observations_for_type.join(
+                            pivoted,
+                            how="left",
+                            on=["response_key", *response_cls.primary_key],
+                        )
+
+                    joined = (
+                        joined.with_columns(
+                            polars.concat_str(
+                                response_cls.primary_key, separator=", "
+                            ).alias(
+                                "__tmp_index_key__"
+                                # Avoid potential collisions w/ primary key
+                            )
+                        )
+                        .drop(response_cls.primary_key)
+                        .rename({"__tmp_index_key__": "index"})
+                    )
 
-            dfs.append(joined)
+                    if not joined.is_empty():
```

Comment:
What happens if it is empty?

Reply:
I think maybe an untested edge case, or somehow I remember it happened for me. It happens if say, realization 0 has no or missing responses matching the observations, then this clause
```
first_columns = joined.select(
                            [
                                "response_key",
                                "index",
                                "observation_key",
                                "observations",
                                "std",
                                str(real),
                            ]
                        )
```
fails. So it was a way of doing it only if the response was non-empty, though maybe too much of an edge case?

Reply:
Been poking around some, and I think it can be removed